lasagne/layers/recurrent.py

```diff
@@ -611,42 +576,49 @@ def step(input_dot_W_n, cell_previous, hid_previous):
             hid = outgate*self.nonlinearity_out(cell)
             return [cell, hid]
 
-        def step_masked(input_dot_W_n, mask, cell_previous, hid_previous):
+        def step_masked(input_dot_w_n, mask_n, cell_previous, hid_previous):
 
-            cell, hid = step(input_dot_W_n, cell_previous, hid_previous)
+            cell, hid = step(input_dot_w_n, cell_previous, hid_previous)
 
             # If mask is 0, use previous state until mask = 1 is found.
             # This propagates the layer initial state when moving backwards
             # until the end of the sequence is found.
-            not_mask = 1 - mask
-            cell = cell*mask + cell_previous*not_mask
-            hid = hid*mask + hid_previous*not_mask
+            not_mask = 1 - mask_n
+            cell = cell*mask_n + cell_previous*not_mask
+            hid = hid*mask_n + hid_previous*not_mask
 
             return [cell, hid]
 
-        if self.backwards and mask is not None:
+        if mask is not None:
             # mask is given as (batch_size, seq_len). Because scan iterates
             # over first dimension, we dimshuffle to (seq_len, batch_size) and
             # add a broadcastable dimension
             mask = mask.dimshuffle(1, 0, 'x')
-            sequences = [input_dot_W, mask]
+            sequences = [input_dot_w, mask]
             step_fun = step_masked
         else:
-            sequences = input_dot_W
+            sequences = input_dot_w
             step_fun = step
 
+        # repeat cell and hid init to batch size
+        ones = T.ones((self.num_batch, 1))
```

Comment:
I dont know i you can do that because the hid_init needs to be the same size as hid inside scan?

Did you see the possible bug with batch size 1?